### PR TITLE
Improve macro call formatting

### PR DESCRIPTION
**Pull request by mglb**
Fixes: #1130 

Improved formatting of macro calls which takes into account many edge cases (like comments in various places of the call).

The new formatting is applied to:
* Standalone macro calls
* Macro calls nested inside standalone macro calls and other affected nested calls

Support for other cases (calls inside `if` conditions, in assignments, etc) will be done after implementing layout optimizer formatting for other structures.
Warning: the reshaping code is pretty messy, and does not work perfectly. However, it works better than current implementation. I'm going to revisit it after refactoring TokenPartitionTree and implementing kChoice partition policy. The choice operator should simplify things a lot.

There are also some other features:
* Partition policies for some Layout Optimizer combinators (juxtaposition, stack, wrap)
* Improvements in layout optimizer's wrapper: token-level wrapping, hanging indentation support
* Few fixes

Some new features in Layout Optimizer (mostly token-level wrapping and hanging indentation in Wrapper function) are not perfect. I''ll improve them when doing other tasks. They will be needed at some point in order to get ternary operator tabular alignment.

Sorry for a huge PR :) It can be safely reviewed commit after a commit.

### CODE REVIEW

**Comment by mglb** [comment]
Sorry for the delay - When I thought everything works, I've generated a few failing test cases that are not really corner cases and had to change a few things in order to make them work.

File: common/formatting/unwrapped_line.h [hunk 1]
```diff
@@ -94,9 +94,13 @@ enum class PartitionPolicyEnum {
   // FormatStyle.wrap_spaces when wrapping.
   kAppendFittingSubPartitions,
 
-  // Does optimizations on partition tree and its subpartitions.
-  // Currently it used for reshaping function calls partitions.
-  kOptimalFunctionCallLayout,
+  // Layout Optimizer policies. See its implementation (layout_optimizer.cc)
```
Comment (fangism) [inline common/formatting/unwrapped_line.h:97]: `LayoutOptimizer` (no space)

Same in other places that reference the class name.

File: common/formatting/unwrapped_line.h [hunk 2]
```diff
@@ -94,9 +94,13 @@ enum class PartitionPolicyEnum {
...
+  // WARNING: when one of the following policies is assigned to a partition,
+  // it's whole subtree is handled by Layout Optimizer.
```
Comment (fangism) [inline common/formatting/unwrapped_line.h:99]: its

File: common/util/iterator_range.h [hunk 3]
```diff
@@ -51,14 +51,19 @@ class iterator_range {
   iterator end_;
 };
 
+template <typename Iter>
+iterator_range(Iter, Iter) -> iterator_range<Iter>;
```
Comment (fangism) [inline common/util/iterator_range.h:55]: Can we double check whether this is supported in C++17 (if that is our current supported standard)?

File: common/util/iterator_range.h [hunk 4]
```diff
@@ -51,14 +51,19 @@ class iterator_range {
...
 // Helper function returning an iterator_range using template argument
 // deduction.
+// OBSOLETE: Use iterator_traits constructor directly.
 template <typename Iter>
 iterator_range<Iter> make_range(Iter begin, Iter end) {
   return iterator_range<Iter>(std::move(begin), std::move(end));
 }
 
 // Overload to operator on a std::pair of iterators.
+// OBSOLETE: Use iterator_traits constructor directly.
```
Comment (fangism) [inline common/util/iterator_range.h:66]: I don't think the above works for this `pair` overload, but check to make sure.

File: verilog/formatting/formatter_test.cc [hunk 5]
```diff
@@ -5499,8 +5498,7 @@ static constexpr FormatterTestCase kFormatterTestCases[] = {
         "      int count;\n"
         "      if (cfg.enable) begin\n"
         "        count = 1;\n"
-        "      end,\n"
-        "      utils_pkg::decrement())\n"
+        "      end, utils_pkg::decrement())\n"
```
Comment (fangism) [inline verilog/formatting/formatter_test.cc]: The previous version might be clearer, depending on whom you ask.  It is a weird case with unlexed code being a macro call argument.

**Review by fangism — CHANGES_REQUESTED** [review]
Preliminary comments.  Haven't finished everything yet, this one was a lot of work!

File: common/formatting/layout_optimizer.cc [hunk 6]
```diff
@@ -182,6 +188,21 @@ LayoutFunction::const_iterator LayoutFunction::AtOrToTheLeftOf(
   return it - 1;
 }
 
+LayoutFunction LayoutFunctionFactory::WrappedLine(
+    const UnwrappedLine& uwline) const {
+  const auto tokens = uwline.TokensRange();
+  auto token_lfs = absl::FixedArray<LayoutFunction>(tokens.size());
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/layout_optimizer.cc:194]: Why? Generally I prefer this way of constructing because it always work (avoids Most Vexing Parse problem, never accidentally calls initializer_list constructor like `Type var{arg, ...};` could), and since C++17 there's guaranteed copy elision in such assignments.

File: common/formatting/layout_optimizer.cc [hunk 7]
```diff
@@ -465,13 +484,123 @@ void TokenPartitionsLayoutOptimizer::Optimize(int indentation,
...
+      break;
+    }
 
-  const auto calculate_optimal_layout_func =
-      std::bind(&TokenPartitionsLayoutOptimizer::CalculateOptimalLayout, this,
-                std::placeholders::_1);
+    case PartitionPolicyEnum::kStack:
+    case PartitionPolicyEnum::kAlwaysExpand:
+    case PartitionPolicyEnum::kTabularAlignment: {
+      int indentation = node.Value().IndentationSpaces();
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/layout_optimizer.cc]: Done

File: common/formatting/layout_optimizer.cc [hunk 8]
```diff
@@ -465,13 +484,123 @@ void TokenPartitionsLayoutOptimizer::Optimize(int indentation,
...
+      std::transform(node.Children().begin(), node.Children().end(),
+                     layouts.begin(), [=](const TokenPartitionTree& n) {
+                       int relative_indentation =
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/layout_optimizer.cc]: Done

File: common/formatting/layout_optimizer.cc [hunk 9]
```diff
@@ -465,13 +484,123 @@ void TokenPartitionsLayoutOptimizer::Optimize(int indentation,
...
 
   switch (node.Value().PartitionPolicy()) {
+    case PartitionPolicyEnum::kJuxtaposition:
+      return factory_.Juxtaposition(layouts.begin(), layouts.end());
+    case PartitionPolicyEnum::kStack:
+      return factory_.Stack(layouts.begin(), layouts.end());
+    case PartitionPolicyEnum::kWrap: {
+      if (VLOG_IS_ON(0) && node.Children().size() > 2) {
+        int second_indentation = node.Children()[1].Value().IndentationSpaces();
+        for (const auto& child : iterator_range(node.Children().begin() + 2,
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/layout_optimizer.cc:562]: I've already added deduction guide for iterator_range: https://github.com/chipsalliance/verible/pull/1172/commits/8d1131176d5eb57c73b74b2fd475dd8b26618cde
Since Verible explicitly requires C++17 should we still care about older standards here?

File: common/formatting/layout_optimizer.cc [hunk 10]
```diff
@@ -465,13 +484,123 @@ void TokenPartitionsLayoutOptimizer::Optimize(int indentation,
...
+                                                node.Children().end())) {
+          if (child.Value().IndentationSpaces() != second_indentation) {
+            VLOG(0) << "Indentations of subpartitions from the second to the "
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/layout_optimizer.cc:565]: It prints the node, which shows indentations of all children.

File: common/formatting/layout_optimizer.cc [hunk 11]
```diff
@@ -465,13 +484,123 @@ void TokenPartitionsLayoutOptimizer::Optimize(int indentation,
...
+      LayoutFunction juxtaposition;
+      const bool juxtaposition_allowed =
+          !std::any_of(layouts.begin() + 1, layouts.end(),
+                       [](const LayoutFunction& lf) { return lf.MustWrap(); });
+      if (juxtaposition_allowed) {
+        juxtaposition = factory_.Juxtaposition(layouts.begin(), layouts.end());
+      }
+
+      int indentation = node.Value().IndentationSpaces();
+      for (int i = 0; i < static_cast<int>(layouts.size()); ++i) {
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/layout_optimizer.cc]: Fixed

File: common/formatting/layout_optimizer.h [hunk 12]
```diff
@@ -36,41 +36,7 @@
...
-//       { (>>>>>>>>>>[...])                                 // - arguments
-//         { (>>>>>>>>>>["%0d %0d\n" ,]) }                   //   - (arg)
-//         { (>>>>>>>>>>[cfg . num_pulses ,]) }              //   - (arg)
-//         { (>>>>>>>>>>[i ) ,]) }                           //   - (arg)
-//       }
-//     }
-//     { (>>>>>>[UVM_DEBUG )]) }                     //   - (arg)
-//   }
-// }
+// Handles formatting of `node` using Layout Optimizer.
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/layout_optimizer.h]: Done

File: common/formatting/layout_optimizer_internal.h [hunk 13]
```diff
@@ -134,7 +134,16 @@ class LayoutItem {
     for (const auto& token : tokens_) {
       // TODO (mglb): support all possible break_decisions
       len += token.before.spaces_required;
-      len += token.Length();
+      if (const auto line_break_pos = token.Text().find('\n');
+          line_break_pos != std::string_view::npos) {
+        // Multiline tokens are not really supported.
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/layout_optimizer_internal.h:139]: Not yet, relevant issue: https://github.com/chipsalliance/verible/issues/1173

File: common/formatting/unwrapped_line.h (hunk 1 above)
Comment (mglb) [reply common/formatting/unwrapped_line.h]: Done

File: common/formatting/unwrapped_line.h (hunk 2 above)
Comment (mglb) [reply common/formatting/unwrapped_line.h]: Fixed

File: common/util/iterator_range.h (hunk 3 above)
Comment (mglb) [reply common/util/iterator_range.h:58]: C++17 final draft: https://timsong-cpp.github.io/cppwp/n4659/temp.deduct.guide
Yes, Verible requires C++17.

File: common/util/iterator_range.h (hunk 4 above)
Comment (mglb) [reply common/util/iterator_range.h]: You are right! I've added missing constructor, deduction guide, and tests.

File: verilog/formatting/formatter_test.cc (hunk 5 above)
Comment (mglb) [reply verilog/formatting/formatter_test.cc:5501]: You are right. Generally argument N+1 should always start a new line when argument N is multiline. A more common case when it happens is just a long argument (some arithmetic expression, function call, etc) that has been wrapped.
This is something I plan to implement in the future. So far its just a TODO comment in the code. I'll make GH issues for such problems soon.

File: verilog/formatting/formatter_test.cc [hunk 14]
```diff
@@ -8334,17 +8332,18 @@ static constexpr FormatterTestCase kFormatterTestCases[] = {
      "  , second\n"
      "} e;\n",
      "typedef enum {\n"
-     "  first  /* c1 */,\n"
-     "  second\n"
+     "    first   /* c1 */\n"
+     "  , second\n"
```
[the remark this answers is not in the document]
Comment (mglb) [reply verilog/formatting/formatter_test.cc:8336]: As a general style preference I agree. However, the input is:
```systemverilog
typedef enum {
  first /* c1 */
  , second
} e;
```
so we deduce that an author wants a leading comma.

This has been requested in: https://github.com/chipsalliance/verible/pull/933#discussion_r711258938

File: common/formatting/layout_optimizer.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (hzeller) [reply common/formatting/layout_optimizer.cc:194]: I guess both ways are fine; the `absl::FixedArray<LayoutFunction> token_lfs(tokens.size());` would be slightly quicker to parse for me as you never have to worry about if copy elision happens or not (but I guess that is only for me as I am working in various code-bases from c++03 to c++17). So I personally always would write it the way David suggested.

But with C++17 in mind, the assignment reads slighly more natural, I don't feel strongly either way. Looks good.

File: common/formatting/layout_optimizer.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (hzeller) [reply common/formatting/layout_optimizer.cc:562]: CTAD was be known to be buggy in various compilers leading up to 17, in particular implicit deduction, but I think these days we can rely on it to be working if we have a deduction guide.

File: common/util/iterator_range.h (hunk 3 above)
Comment (hzeller) [reply common/util/iterator_range.h:58]: We also compile with `--cxxopt=-Wctad-maybe-unsupported` to check for implicit deduction without guide, so since this deduction guide is added here, that should work, and the CI will tell.

(What we should do, in a separate cleanup PR, is to use `iterator_range()` everywhere and remove `make_range()` to minimize confusion about what to use)

File: verilog/formatting/formatter_test.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (hzeller) [reply verilog/formatting/formatter_test.cc:8336]: Yes, this is now working as intended. We got several issues from users that use this style in their code and found that Verible verschlimmbesserte their code.

File: verilog/formatting/formatter_test.cc (hunk 5 above)
Comment (mglb) [reply verilog/formatting/formatter_test.cc:5501]: Done https://github.com/chipsalliance/verible/issues/1211

File: common/formatting/layout_optimizer.cc [hunk 15]
```diff
@@ -465,13 +484,123 @@ void TokenPartitionsLayoutOptimizer::Optimize(int indentation,
...
+                                                node.Children().end())) {
+          if (child.Value().IndentationSpaces() != second_indentation) {
+            VLOG(0) << "Indentations of subpartitions from the second to the "
+                       "last are not equal. Using indentation of the second "
+                       "subpartition as a hanging indentation. Parent node:\n"
+                    << node;
+          }
+        }
+      }
+      int hanging_indentation =
```
[the remark this answers is not in the document]
Comment (mglb) [reply common/formatting/layout_optimizer.cc]: Done

**Comment by hzeller** [comment]
Thanks, merged!
